boa_parser/src/source.rs

```diff
@@ -0,0 +1,88 @@
+use std::{
+    fs::File,
+    io::{self, BufReader, Read},
+    path::Path,
+};
+
+/// A source of ECMAScript code.
+///
+/// [`Source`]s can be created from plain [`str`]s, file [`Path`]s or more generally, any [`Read`]
+/// instance.
+#[derive(Debug)]
+pub struct Source<'path, R> {
+    pub(crate) reader: R,
+    pub(crate) path: Option<&'path Path>,
+}
+
+impl<'bytes> Source<'static, &'bytes [u8]> {
+    /// Creates a new `Source` from any type equivalent to a slice of bytes e.g. [`&str`], [`Vec<u8>`]
+    /// <code>[Box]<[\[u8\]][slice]></code> or a plain slice <code>[&\[u8\]][slice]</code>.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use boa_parser::Source;
+    /// let code = r#"var array = [5, 4, 3, 2, 1];"#;
+    /// let source = Source::from_bytes(code);
+    /// ```
+    ///
+    /// [`&str`]: str
+    /// [slice]: std::slice
```

Comment:
```suggestion
    /// Creates a new `Source` from any type equivalent to a slice of bytes e.g. [`&str`][str],
    /// <code>[Vec]<[u8]></code>,
    /// <code>[Box]<[\[u8\]][std::slice]></code>
    /// or a plain slice <code>[&\[u8\]][std::slice]</code>.
    ///
    /// # Examples
    ///
    /// ```
    /// # use boa_parser::Source;
    /// let code = r#"var array = [5, 4, 3, 2, 1];"#;
    /// let source = Source::from_bytes(code);
    /// ```
```